### ISSUE TITLE
MVC Example in Readme doesn't work

### ISSUE THREAD

**Issue by 26chairs** [kenkendk/ceenhttpd#19]
Hi - ceen seems like exactly what I'm looking for, but I ran into a few issues trying to set it up. I basically copy/pasted the MVC example from the readme and the Index with no parameters works, but none of the other ones ("/api/v1/{id}", "/api/v1/{id}/detail") work - one gives a 404 and the other a 500.

Is it possible the readme example wasn't updated after a change that modified the way that stuff would need to be set up at some point? Am I missing better documentation that exists elsewhere? I'm about to download the source and debug through it, but wanted to ask as well just in case.

**Comment by kenkendk** [comment]
It is possible that there is something in the README that is out-of-date, but it should work.
There is a unittest here that works for me, testing most of the functionality of the MVC mapping:
https://github.com/kenkendk/ceenhttpd/blob/master/Unittests/MvcTests.cs

That said, there are two cases where stuff can go wrong. One is routing. Since the url's are derived from the classes, I often find myself calling another url than what is generated. This is a result of trying to do something automatically, which results in it being opaque. You can toggle the debug flag in the route controller configuration, and then it should show you all the urls and their mappings in the console output. This should reveal what causes the 404 error.

I am working on an alternate way to doing the mapping, where you can manually specify routes, using the url paths as strings, making the mapping more explicit and transparent. I find that this works better for cases where the API is exposed and needs to stay fixed, and not change when items are renamed etc.

The other place it can go wrong is when attempting to match the request to the handler. The 500 error code suggests that there is an exception happening somewhere.To figure out where this happens, you can add the StdOutErrors logger, which will output all exceptions to the console when running.

I have had another similar report for a 500 error that was caused by not loading `NewtonSoft.Json v12.0.1`. I think there is a missing dependency in the Nuget package.

**Comment by 26chairs** [comment]
Yep, installing the two Ceen packages from Nuget seems to have pulled down the wrong version of Newtonsoft.Json. After updating that dependency, the 500 went away.

The 404 still remains on http://127.0.0.1:8080/api/v1/5 though. Running the web server with the example code from the readme gives me the following on startup:

All target paths:
/api/v1 => Test.ApiExampleController, Ceen.Mvc.IResult Index(Ceen.IHttpContext)
/api/v1 => Test.ApiExampleController, Ceen.Mvc.IResult Index(Int32)
/api/v1/{id}/detail => Test.ApiExampleController, Ceen.Mvc.IResult Detail(Int32)
Map structure:
/api/v1 => Test.ApiExampleController, Ceen.Mvc.IResult Index(Int32)
/api/v1 => Test.ApiExampleController, Ceen.Mvc.IResult Index(Ceen.IHttpContext)
/api/v1/{id}/detail => Test.ApiExampleController, Ceen.Mvc.IResult Detail(Int32)
Serving files, press enter to stop ...

Only way I can seem to get the behavior the example code claims I should get is by throwing a
[Route("{id}")]
on the Index method that takes in an int parameter.

I think I've got what I need though. Thanks a ton for creating and maintaining this project. This is exactly what I was looking for and I couldn't be happier that I stumbled upon it!

**Comment by kenkendk** [comment]
> Only way I can seem to get the behavior the example code claims I should get is by throwing a [Route("{id}")]
> on the Index method that takes in an int parameter.

That is, IMO, how it should work. The path matcher checks that all parts of the path are mapped. For an url path such as `/api/v1/5` or `/api/v1/abc`, the last part is not matching anything, thus you get 404. You could call `/api/v1?id=5` which should hit the right method.
When you add `[Route("{id}")]`, you can see that the generated url is now `/api/v1/{id}` which can now map the path argument `{id}` to the function argument.

This is a good example of why I think the automatic mapping is not always as great as it looks, because it is not completely transparent how the mapping is applied.

My new methods are almost ready, and then you can explicitly map the urls:
```
.WireWith<ApiExampleController>()
	.Wire("GET /api/v1/entry", nameof(ApiExampleController.Index), typeof(IHttpContext))
	.Wire("GET /api/v1/entry/{id}", nameof(ApiExampleController.Index), typeof(IHttpContext), typeof(int), typeof(string))
	.Wire("GET /api/v1/entry/{id}/detail", nameof(ApiExampleController.Cross))
	.Wire("POST /api/v1/entry/{id}", nameof(ApiExampleController.Update))
	.Wire("GET /api/v1/entry/detail", nameof(ApiExampleController.Detail), typeof(IHttpContext))
	.Wire("GET|POST /api/v1/entry/detail/{id}", nameof(ApiExampleController.Detail), typeof(IHttpContext), typeof(int))
```

**Comment by 26chairs** [comment]
Makes sense. I'll keep an eye out for updates to the repository.

Thanks for the help! Really appreciate your work - was able to throw together a quick REST API in minutes.

**Comment by kenkendk** [comment]
I have updated the code just now, and released new NuGet packages fo `v0.9.9` that fixes the json dependency and adds the manual wiring options.
